src/providers/mssql/qgsmssqltransaction.cpp

```diff
@@ -0,0 +1,158 @@
+/***************************************************************************
+  qgsmssqltransaction.cpp - QgsMssqlTransaction
+ ---------------------
+ begin                : 11.3.2021
+ copyright            : (C) 2021 by Vincent Cloarec
+ email                : vcloarec at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmssqltransaction.h"
+
+#include "qgsmssqldatabase.h"
+
+#include "qgsexpression.h"
+#include "qgsmessagelog.h"
+
+#include <QSqlError>
+#include <QSqlQuery>
+#include <QtDebug>
+
+
+QgsMssqlTransaction::QgsMssqlTransaction( const QString &connString )
+  : QgsTransaction( connString )
+{
+}
+
+QgsMssqlTransaction::~QgsMssqlTransaction()
+{
+}
+
+
+bool QgsMssqlTransaction::executeSql( const QString &sql, QString &error, bool isDirty, const QString &name )
+{
+  QSqlDatabase &database = mConn->db();
+
+  if ( !database.isValid() )
+    return false;
+  if ( !database.isOpen() )
+    return false;
+
+  if ( isDirty )
+  {
+    QgsTransaction::createSavepoint( error );
+    if ( ! error.isEmpty() )
+    {
+      return false;
+    }
+  }
+
+  QSqlQuery query( database );
+  if ( !query.exec( sql ) )
+  {
+    if ( isDirty )
+      rollbackToSavepoint( savePoints().last(), error );
+
+    QString msg = tr( "MSSQL query failed: %1" ).arg( query.lastError().text() );
+    if ( error.isEmpty() )
+      error = msg;
+    else
+      error = QStringLiteral( "%1\n%2" ).arg( error, msg );
+
+    return false;
+  }
+
+  if ( isDirty )
+  {
+    dirtyLastSavePoint();
+    emit dirtied( sql, name );
+  }
+
+  return true;
+}
+
+
+QString QgsMssqlTransaction::createSavepoint( const QString &savePointId, QString &error )
+{
+  if ( !mTransactionActive )
+    return QString();
+
+  if ( !executeSql( QStringLiteral( "SAVE TRAN %1" ).arg( QgsExpression::quotedColumnRef( savePointId ) ), error ) )
```

Comment:
so only difference between PG and MSSQL is that it uses different keyword? thinking if it is worth to have all other logic in the function doubled or to have just have this SQL statement or SQL keyword as virtual function in providers?

Reply:
Yeah that's right, just different keyword. I was half-minded here and thought this is the simplest solution. There is similar situation also with other calls like commitTransaction(), rollbackTransaction() where there's just a different keyword used among different providers. But given that the logic is quite simple, this seemed like the least intrusive approach. And in future if some provider supports transactions, but maybe needs multiple queries to be executed, this is more flexible approach...

Reply:
do not have strong opinon, just asking. but for example I do not have MSSQL provider ON in my dev setup, so it could happen that if I just modify the qgstransaction function I would not find out that there is this alternative implementation here. But I am happy with this way too.

Reply:
cool... by the way, MSSQL provider is always ON :-) the main issue really is that currently MSSQL tests are not run in CI because of the problems Nyall mentioned above - that's what really needs fixing to make sure things do not break...